### PR TITLE
Multi_comment

### PR DESCRIPTION
**Pull request by Bioamin**
include updates for parallel processing.

### CODE REVIEW

**Comment by tlawrence3** [comment]
* `ilm.py` 
   * `lines 166-174`
Currently, initial parameter values are printed once for each worker, which is not the desired behavior. If you move the functionality of this code to before `line 197` it will only be printed once restoring the original intention.

    * `line 176`
You should remove this line because its output will be incoherent, because each worker will have its own set of trails and iterations. This functionality will be included in the for loops starting on `line 207`.

**Comment by tlawrence3** [comment]
> putting if before result_dict[trial][gen][obs] cause the code not to print anything since we always have a None parameter in the dictionary. for example, when we don't want to print the matrix but we do want to print the functional load. It does not print anything because the Matrix parameter is None.
maybe a for loop can fix this problem.

@Bioamin don't worry about that part just yet. Let's address my previous comments first since they will be easier to understand and fix.

**Review by tlawrence3 — CHANGES_REQUESTED** [review]
The new `if` statement needs to be in the `for obs` loop not outside it. This will also print the whole result dictionary multiple times instead of a single line for each generation.

**Review by tlawrence3 — CHANGES_REQUESTED** [review]
* `ilm.py` 
   * `lines 166-174`
Currently, initial parameter values are printed once for each worker, which is not the desired behavior. If you move the functionality of this code to before `line 197` it will only be printed once restoring the original intention.

    * `line 176`
You should remove this line because its output will be incoherent, because each worker will have its own set of trails and iterations. This functionality will be included in the for loops starting on `line 207`.

**Comment by Bioamin** [comment]
The new error occurred. which said in pool.py 
 'dict' object is not callable

**Review by tlawrence3 — CHANGES_REQUESTED** [review]
* `Line 182`
You shouldn't need to return the matrix learner object (parent). I know your reasoning behind this, however, you don't need access to the `parent` object to produce the desired output for `lines 198-205`.

* `Lines 198-205`
You need to modify these lines to produce the desired output. Specifically, you need to change `line 198` and `line 205` to sever the dependency on the `matrix learner` (parent) object. On quick inspection it seems that the `print_parameters()` method just prints initial values of model parameters which are also contained in the arguments parsed from the command line so you don't actually need the object to recreate this output. The `print_observables_header()` just produces an output header line, which you can recreate without needing the object.  

* `Line 185` I am not sure what the purpose of this line of code is? It is not needed and should be removed.

File: bin/ilm [hunk 1]
```diff
@@ -147,25 +154,78 @@ ilm "(([a-g]\[aeiou]):0.1)^2"   "{256}.(2)"  # any sound space can be powered
...
+#                    print('# Trial {} Iteration {}'.format(trial,generation))
+                    child = parent.spawn()
+                    lessons = parent.teach(options.num_interactions)
+                    child.learn(lessons)
+                    result_dict[trial][generation]=child.print_observables()
+                    parent = child
+        return(result_dict,parent)
+
+
+    result_dict,parent=multidef (num_trials,options)
```
Comment (Bioamin) [inline bin/ilm:185]: If I remove line 185 I will get the following error:
    result=pool.starmap(result_dict,iter_jobs)         #(multidef,iter_jobs)
NameError: name 'result_dict' is not defined

**Comment by Bioamin** [comment]
@tlawrence3 could you give me a hint on how I modify line 198 and 205 in order to have the same result without using "parent"? 

Even when I comment out both of those lines(198 and 205) I get the following error: 


Traceback (most recent call last):
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 119, in worker
    result = (True, func(*args, **kwds))
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 47, in starmapstar
    return list(itertools.starmap(args[0], args[1]))
TypeError: 'dict' object is not callable
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/aboroomand/.conda/envs/ILMenv/bin/ilm", line 4, in <module>
    __import__('pkg_resources').run_script('ILMpy==0.1', 'ilm')
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 748, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/pkg_resources/__init__.py", line 1517, in run_script
    exec(code, namespace, namespace)
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/ILMpy-0.1-py3.6.egg/EGG-INFO/scripts/ilm", line 206, in <module>
    result=pool.starmap(result_dict,iter_jobs)         #(multidef,iter_jobs)
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 274, in starmap
    return self._map_async(func, iterable, starmapstar, chunksize).get()
  File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 644, in get
    raise self._value
TypeError: 'dict' object is not callable

**Comment by dhard** [comment]
Hi Amin,

Can you make swifter progress in your iterative progress of new results for your ILM project?

Please don’t be discouraged. If you need my help regarding access to MERCED cluster, please let me know.

Cheers
Dave

Sent from my iPhone

On Nov 28, 2017, at 10:20, Amin Boroomand <notifications@github.com<mailto:notifications@github.com>> wrote:


@tlawrence3<https://github.com/tlawrence3> could you give me a hint on how I modify line 198 and 205 in order to have the same result without using "parent"?

Even when I comment out both of those lines(198 and 205) I get the following error:

Traceback (most recent call last):
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 119, in worker
result = (True, func(*args, **kwds))
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 47, in starmapstar
return list(itertools.starmap(args[0], args[1]))
TypeError: 'dict' object is not callable
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
File "/home/aboroomand/.conda/envs/ILMenv/bin/ilm", line 4, in
import('pkg_resources').run_script('ILMpy==0.1', 'ilm')
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/pkg_resources/init.py", line 748, in run_script
self.require(requires)[0].run_script(script_name, ns)
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/pkg_resources/init.py", line 1517, in run_script
exec(code, namespace, namespace)
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/site-packages/ILMpy-0.1-py3.6.egg/EGG-INFO/scripts/ilm", line 206, in
result=pool.starmap(result_dict,iter_jobs) #(multidef,iter_jobs)
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 274, in starmap
return self._map_async(func, iterable, starmapstar, chunksize).get()
File "/home/aboroomand/.conda/envs/ILMenv/lib/python3.6/multiprocessing/pool.py", line 644, in get
raise self._value
TypeError: 'dict' object is not callable

**Pull request closed without merging**